Make function list more compact

light-and-salt opened the pull request:
Fix #196

Show "above the fold" functions (aggregations, bins, timeUnits) in the `functionselect` panel. Functions below the fold will reveal if user clicks `more`.

The order of the above the fold functions is specified in #196. 

Used local aggregation and timeUnit list (instead of getting from schema) because aggregations and timeUnits are unlikely to change.

Count fields has only one function: `COUNT`

Passed `npm run lint`

Passed `npm run test`

![q](https://cloud.githubusercontent.com/assets/822034/16883861/242e07da-4a7b-11e6-90fd-f7035827954e.gif)

![t](https://cloud.githubusercontent.com/assets/822034/16883866/274ac912-4a7b-11e6-9477-5e38001b9188.gif)

light-and-salt committed:
Make function list more compact

kanitw commented:
The order is correct, but the layout is not correct yet.   (Put None alone on the first line, make "None" "gray").

kanitw commented on src/components/functionselect/functionselect.js:
```diff
@@ -14,25 +14,50 @@ angular.module('vlui')
-        function getFns(type) {
+        // timeUnits for T
+        var timeUnits = {
+          aboveFold: ["year", "quarter", "month", "date", "day", "hours", "yearmonthdate"],
```
move minutes and seconds up here.

// TODO(later): use statistics to determine what's above fold and below fold (Waiting for CompassQL schema to provide this info)

light-and-salt replied:
Done

kanitw commented on src/components/functionselect/functionselect.js:
```diff
@@ -14,25 +14,50 @@ angular.module('vlui')
+        // timeUnits for T
+        var timeUnits = {
+          aboveFold: ["year", "quarter", "month", "date", "day", "hours", "yearmonthdate"],
+          belowFold: ["hoursminutes", "hoursminutesseconds", "milliseconds", 
```
Remove yeardate, yearday -- they are bad.  I'm removing them from Vega-Lite.  

yearmonthday => yearmonthdate  (same for longer version of this)

remove yearquartermonth, quarter month to reduce # of redundant options.

light-and-salt replied:
I noticed that vega-lite doesn't have `yearmonthdatehours`, `yearmonthdatehoursminutes`, `yearmonthdatehoursminutesseconds`, but vega-lite has `yearmonthdayhours` etc.

To me, `2016-July-15-13:50:00` (`yearmonthdatehousminutesseconds`) makes more sense than `2016-July-Friday-13:50:00` (`yearmonthdayhoursminutesseconds`). `day` may be useful for seeing weekly patterns, but `date` is more accurate (e.g., there are multiple Fridays in July but there is only one 15 in July).

light-and-salt committed:
Improve function list CSS

light-and-salt commented:
@kanitw This PR is ready for review again. I've addressed your comments and improved the function list CSS. Here's how it looks like right now:

![screen shot 2016-07-15 at 3 15 19 pm](https://cloud.githubusercontent.com/assets/822034/16889947/2527ff28-4a9f-11e6-8af3-9ddb0de336f1.png)

![screen shot 2016-07-15 at 3 15 01 pm](https://cloud.githubusercontent.com/assets/822034/16889944/21be342e-4a9f-11e6-90dc-fa3bbf5cafba.png)

kanitw commented:
Umm why 3 columns?  My mock two columns format group similar functions together on the same row (min and max, mean and median).  Is there any reason for 3?

TimeUnit's column are not aligned.

kanitw commented:
(3 columns might be good for timeUnit though, as there are too many)

Are `argmax` and `argmin` and `values` sensible here?  Maybe not?

light-and-salt commented:
Oh I made the panel wider because some timeUnit are really long (besides there are many timeUnits). I can make the Q function list two columns. Just a sec...

> Are `argmax` and `argmin` and `values` sensible here?  Maybe not?

Right, I'll remove these

light-and-salt commented:
I've changed some CSS

Above-the-fold functions are laid out in two columns:
![screen shot 2016-07-15 at 5 16 27 pm](https://cloud.githubusercontent.com/assets/822034/16891595/2bd834f2-4ab1-11e6-8502-ef06c390fe9e.png)

![screen shot 2016-07-15 at 5 16 58 pm](https://cloud.githubusercontent.com/assets/822034/16891598/31d2c3f4-4ab1-11e6-836c-4f10f1836178.png)

Below-the-fold functions are laid out in a single-column, because they are sorted alphabetically:

![screen shot 2016-07-15 at 5 16 45 pm](https://cloud.githubusercontent.com/assets/822034/16891607/46a9fd92-4ab1-11e6-926e-bcc0d30fda1b.png)

![screen shot 2016-07-15 at 5 17 13 pm](https://cloud.githubusercontent.com/assets/822034/16891608/4be97bca-4ab1-11e6-86a1-8dcb5a9a231d.png)

light-and-salt committed:
Use `date` timeUnits

light-and-salt commented:
This PR is ready for review @kanitw 

![screen shot 2016-07-20 at 11 41 03 pm](https://cloud.githubusercontent.com/assets/822034/17013475/acbdd998-4ed3-11e6-9391-3d67340b5783.png)

kanitw commented on src/components/functionselect/functionselect.js:
```diff
@@ -14,25 +14,49 @@ angular.module('vlui')
+        function getTimeUnits(type) {
           if (type === 'temporal') {
-            return Schema.schema.definitions.TimeUnit.enum;
+            if (!timeUnits.all || timeUnits.all.length <= 0) {
```
I don't understand why you have to do lazy initialization for `timeUnits.all`  why don't you just initialize it right after when you define `timeUnits`?

kanitw commented on src/components/functionselect/functionselect.js:
```diff
@@ -14,25 +14,49 @@ angular.module('vlui')
           // TODO: make this correct for temporal as well
           if (type === 'quantitative' ){
-            return Schema.schema.definitions.AggregateOp.enum;
+            if (!aggregations.all || aggregations.all.length <= 0) {
```
I don't understand why you have to do lazy initialization for `aggregations.all`  why don't you just initialize it right after when you define `aggregations`?

Plus, please rename this to `aggregates` for consistency.

light-and-salt replied:
Done

kanitw commented:
- [x] These changes do not pass lint.

kanitw committed:
Partially cleaning up functionselection for @zeningqu

- Reformat the `aboveFold` and `belowFold` array to better reflects its appearance in the UI.
- Remove `average` as it is redundant with `mean`  (Why include both?)
- Use ' instead of "
- Regroup stuff belowFold for aggregation.   @zeningqu please make "below fold" for aggregations two columns.

kanitw commented:
- [x] I don't think belowFold for aggregations should be one column.

kanitw commented on src/components/functionselect/functionselect.js:
```diff
@@ -14,25 +14,75 @@ angular.module('vlui')
+          belowFold: [
+            'hoursminutes',
+            'hoursminutesseconds',
+            'milliseconds',
```
Why does milliseconds get mixed up with multi-part time units?

light-and-salt committed:
Re-order timeUnits in function list

kanitw commented on src/components/fieldinfo/fieldinfo.scss:
```diff
@@ -16,6 +16,16 @@ $any-hover-color: #666;
   user-select: none;
   overflow: hidden;
 
+  .remove {
```
Can you remove this commit (ba503da51dabc1e17826a7bf5f5cc0f40f304df9) entirely? 

I don't think this PR address the problem is a correct way.  It introduces so many problems.
- Position absolute isn't the right solution for flex box
- max-width: 100px is way too small.  

If you want to address this problem, file an issue and address it in a separate PR.  (It's not the crux of this PR.)

light-and-salt replied:
Removed

light-and-salt committed:
Rename aggregations to aggregates for consistency

light-and-salt committed:
Make belowFold timeUnits single-column and belowFold aggregates double-column

light-and-salt committed:
Hide "more" & "less" toggles for COUNT

light-and-salt committed:
Modify test because function list has changed

light-and-salt committed:
Clean up

kanitw commented on src/components/functionselect/functionselect.js:
```diff
@@ -14,25 +14,72 @@ angular.module('vlui')
+            'variance', 'variancep'
+          ] // there is no 'count' for Q
+        };
+        aggregates.all = aggregates.aboveFold.concat(aggregates.belowFold);
```
Line 105 depends on `aggregates.all` but your aggregates.all doesn't include count.

light-and-salt replied:
Thank you!!! !!!

light-and-salt replied:
This is exactly why `$watch('fieldDef')` is triggered a second time without aggregation!

kanitw commented:
- [x] No matter what, please reverse a71dfb9.

light-and-salt committed:
Fix $watch(fieldDef) trigger loop bug. Thanks to Ham!

kanitw commented on src/components/functionselect/functionselect.js:
```diff
@@ -112,6 +112,8 @@ angular.module('vlui')
 
         // when parent objects modify the field
         scope.$watch('fieldDef', function(pill) {
+          console.log('fieldDef changed: ', scope.fieldDef);
```
log...

kanitw commented on src/components/functionselect/functionselect.js:
```diff
@@ -14,25 +14,72 @@ angular.module('vlui')
 
-        function getAggrs(type) {
+        function getAggregates(type) {
           if(!type) {
```
Why !type?

kanitw commented on src/components/functionselect/functionselect.js:
```diff
@@ -55,8 +102,8 @@ angular.module('vlui')
           pill.bin = selectedFunc === BIN ? true : undefined;
-          pill.aggregate = getAggrs(type).indexOf(selectedFunc) !== -1 ? selectedFunc : undefined;
-          pill.timeUnit = getFns(type).indexOf(selectedFunc) !== -1 ? selectedFunc : undefined;
+          pill.aggregate = getAggregates(type).concat(['count']).indexOf(selectedFunc) !== -1 ? selectedFunc : undefined;
```
This concat is very weird.  You only use aggregate.all  once, why don't just you add count to `aggregates.all`?   (Add comment explain why there too.)

light-and-salt committed:
Return no aggregates for undefined type

kanitw commented:
- [x] merge conflicts
- [x] clean up a bit more

light-and-salt committed:
Remove Schema because it's no longer used in functionSelect

light-and-salt committed:
Use average instead of mean, based on Jeff's feedback

light-and-salt committed:
Say Function & Type instead of Functions & Types, based on Jeff's feedback

light-and-salt committed:
Separate bin with other aggregates, and re-order aggregates a bit

light-and-salt committed:
Re-order functions

light-and-salt commented:
The latest screenshots:

![screen shot 2016-08-12 at 2 01 28 pm](https://cloud.githubusercontent.com/assets/822034/17636951/5651e440-6095-11e6-99e5-e593a246f9b2.png)

![screen shot 2016-08-12 at 2 01 36 pm](https://cloud.githubusercontent.com/assets/822034/17636957/59b2abe2-6095-11e6-8fbf-c640867e3053.png)

![screen shot 2016-08-12 at 2 02 41 pm](https://cloud.githubusercontent.com/assets/822034/17636979/7d99acc2-6095-11e6-978b-190820ae8d56.png)

![screen shot 2016-08-12 at 2 02 49 pm](https://cloud.githubusercontent.com/assets/822034/17636986/811b1944-6095-11e6-991d-2f3557ba2af4.png)

light-and-salt committed:
Explain in comment that we want bin for quantitative fields

kanitw committed:
Remove weird tabbing 

cc: @zeningqu

kanitw committed:
// HACK
